File: server/src/routes/ssh.ts
import type { FastifyInstance } from 'fastify';
import { z } from 'zod';
import type {
  HostPreviewResponse,
  OpenSshMetadataPatch,
  OpenSshProfileMetadata,
  SshConfigResponse,
  SshKeysResponse,
} from '@muxus/shared';
import type { AppContext } from '../app.js';
import { sendError } from '../util/errors.js';
import { metadataPatchSchema } from './metadata-schema.js';
import { defaultSshConfigPath, listHosts, loadConfigDocument } from '../ssh/ssh-config.js';
import { deleteHost, previewHost, upsertHost } from '../ssh/ssh-config-edit.js';
import { listSshKeys } from '../ssh/key-scan.js';

const forwardSchema = z.object({
  type: z.enum(['local', 'remote', 'dynamic']),
  bindPort: z.number().int().min(1).max(65535),
  targetHost: z.string().min(1).optional(),
  targetPort: z.number().int().min(1).max(65535).optional(),
});

const upsertSchema = z.object({
  aliases: z.array(z.string().min(1)).min(1),
  description: z.string().max(2000).optional(),
  file: z.string().optional(),
  previousAlias: z.string().optional(),
  options: z.object({
    hostname: z.string().optional(),
    user: z.string().optional(),
    port: z.number().int().min(1).max(65535).optional(),
    identityFiles: z.array(z.string()).optional(),
    identitiesOnly: z.boolean().optional(),
    forwardAgent: z.boolean().optional(),
    proxyJump: z.array(z.string()).optional(),
    forwards: z.array(forwardSchema).optional(),
    passwordOnly: z.boolean().optional(),
    extras: z.array(z.object({ keyword: z.string(), value: z.string() })).optional(),
  }),
});

/** Live OpenSSH config plus Muxus-owned metadata, editing, and key discovery. */
export function registerSshRoutes(app: FastifyInstance, ctx: AppContext): void {
  app.get('/api/ssh/config', (): SshConfigResponse => {
    const doc = loadConfigDocument();
    const hosts = listHosts(doc).sort((a, b) => a.alias.localeCompare(b.alias));
    const metadata = ctx.database.openSshMetadata(hosts.map((host) => host.alias));
    return {
      path: defaultSshConfigPath(),
      files: doc.fileOrder,
      hosts: hosts.map((host) => ({ ...host, metadata: metadata.get(host.alias) })),
      error: doc.error,
    };
  });

  app.post('/api/ssh/config/hosts', async (req, reply) => {
    try {
      const parsed = upsertSchema.safeParse(req.body);
      if (!parsed.success) {
        return await reply.code(400).send({ message: parsed.error.issues[0]?.message ?? 'invalid host payload' });
      }
      const result = upsertHost(parsed.data);
      const nextAlias = parsed.data.aliases[0]!;
      if (parsed.data.previousAlias) {
        ctx.database.renameOpenSshAlias(parsed.data.previousAlias, nextAlias);
      }
      return result;
    } catch (err) {
      return sendError(reply, err);
    }
  });

  app.post('/api/ssh/config/preview', async (req, reply): Promise<HostPreviewResponse | void> => {
    try {
      const parsed = upsertSchema.safeParse(req.body);
      if (!parsed.success) {
        return await reply.code(400).send({ message: parsed.error.issues[0]?.message ?? 'invalid host payload' });
      }
      return { text: previewHost(parsed.data) };
    } catch (err) {
      return sendError(reply, err);
    }
  });

  app.delete('/api/ssh/config/hosts/:alias', async (req, reply) => {
    try {
      const { alias } = req.params as { alias: string };
      deleteHost(alias);
      return { ok: true };
    } catch (err) {
      return sendError(reply, err);
    }
  });

  app.patch('/api/ssh/config/hosts/:alias/metadata', async (req, reply): Promise<OpenSshProfileMetadata | void> => {
    try {
      const { alias } = req.params as { alias: string };
      const parsed = metadataPatchSchema.safeParse(req.body);
      if (!parsed.success) {
        return await reply.code(400).send({ message: parsed.error.issues[0]?.message ?? 'invalid metadata payload' });
      }
      return ctx.database.updateOpenSshMetadata(alias, parsed.data satisfies OpenSshMetadataPatch);
    } catch (err) {
      return sendError(reply, err);
    }
  });

  app.get('/api/ssh/keys', async (): Promise<SshKeysResponse> => listSshKeys());
}

File: app/internal/server/routes/ssh.go
package routes

import (
	"encoding/json"
	"net/http"
	"sort"

	"github.com/go-chi/chi/v5"

	"github.com/FloSch62/muxus/app/internal/api"
	"github.com/FloSch62/muxus/app/internal/persist"
	"github.com/FloSch62/muxus/app/internal/server"
	"github.com/FloSch62/muxus/app/internal/sshx"
)

func writeConfigError(w http.ResponseWriter, err error) {
	if problem, ok := err.(*sshx.ConfigProblem); ok {
		server.WriteJSON(w, problem.Status, api.ErrorBody{
			Message: problem.Message,
			Code:    problem.Code,
		})
		return
	}
	writeInternalError(w, err)
}

func RegisterSSHRoutes(r chi.Router, db *persist.DB) {
	r.Get("/api/ssh/config", func(w http.ResponseWriter, _ *http.Request) {
		path := sshx.DefaultConfigPath()
		doc := sshx.LoadConfigDocument(path)
		hosts := sshx.ListHosts(doc)
		sort.SliceStable(hosts, func(i, j int) bool { return hosts[i].Alias < hosts[j].Alias })
		aliases := make([]string, len(hosts))
		for index := range hosts {
			aliases[index] = hosts[index].Alias
		}
		metadata, err := db.OpenSSHMetadata(aliases)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		for index := range hosts {
			if value, ok := metadata[hosts[index].Alias]; ok {
				value := value
				hosts[index].Metadata = &value
			}
		}
		server.WriteJSON(w, http.StatusOK, api.SSHConfigResponse{
			Path: path, Files: doc.FileOrder, Hosts: hosts, Error: doc.Error,
		})
	})

	parseUpsert := func(w http.ResponseWriter, req *http.Request) (api.HostUpsertRequest, bool) {
		var input api.HostUpsertRequest
		body, ok := readJSONBody(w, req, defaultBodyLimit)
		if !ok {
			return input, false
		}
		if err := json.Unmarshal(body, &input); err != nil {
			writeBadRequest(w, "invalid host payload")
			return input, false
		}
		return input, true
	}

	r.Post("/api/ssh/config/hosts", func(w http.ResponseWriter, req *http.Request) {
		input, ok := parseUpsert(w, req)
		if !ok {
			return
		}
		result, err := sshx.UpsertHost(input, sshx.DefaultConfigPath())
		if err != nil {
			writeConfigError(w, err)
			return
		}
		if input.PreviousAlias != "" {
			if err := db.RenameOpenSSHAlias(input.PreviousAlias, input.Aliases[0]); err != nil {
				writeInternalError(w, err)
				return
			}
		}
		server.WriteJSON(w, http.StatusOK, result)
	})

	r.Post("/api/ssh/config/preview", func(w http.ResponseWriter, req *http.Request) {
		input, ok := parseUpsert(w, req)
		if !ok {
			return
		}
		text, err := sshx.PreviewHost(input, sshx.DefaultConfigPath())
		if err != nil {
			writeConfigError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, api.HostPreviewResponse{Text: text})
	})

	r.Delete("/api/ssh/config/hosts/{alias}", func(w http.ResponseWriter, req *http.Request) {
		alias := chi.URLParam(req, "alias")
		if err := sshx.DeleteHost(alias, sshx.DefaultConfigPath()); err != nil {
			writeConfigError(w, err)
			return
		}
		if _, err := db.DeleteSessionLoggingPolicy("ssh:" + alias); err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, map[string]bool{"ok": true})
	})

	r.Patch("/api/ssh/config/hosts/{alias}/metadata", func(w http.ResponseWriter, req *http.Request) {
		body, ok := readJSONBody(w, req, defaultBodyLimit)
		if !ok {
			return
		}
		patch, err := parseMetadataPatch(body)
		if err != nil {
			writeBadRequest(w, err.Error())
			return
		}
		record, err := db.UpdateOpenSSHMetadata(chi.URLParam(req, "alias"), patch)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, record)
	})

	r.Get("/api/ssh/keys", func(w http.ResponseWriter, _ *http.Request) {
		server.WriteJSON(w, http.StatusOK, sshx.ListSshKeys(""))
	})
}
